### PR TITLE
remove deprecated `Drain` field from `structs.Node`

### PR DESCRIPTION
**Pull request by cgbaker**
**Note**: upgrade testing on this PR should verify that `.SecretID` is not emitted by the API (using curl, not the Nomad CLI), and that `.Drain` is emitted.

This is an attempt to remove the `Drain` field from `structs.Node`, while still keeping backwards compatibility with the `api.Node.Drain` field. The questions at play are: 
* do we want to actually delete this field?
* do we like this JSON encoding extensions approach for API backwards compatibility?

Here's what the PR does:
* Remove deprecated `Drain` field from `structs/Node` and deleted code/tests associated with this field.
* Adds encoding extensions to put the `Node.Drain` field back to API responses
* Modifies the HTTP agent and the event stream to use encoders with these extensions
* Also removed special code for emptying `structs.Node.SecretID`, instead allowing the encoder to remove it using a `json` tag on the struct, which works because (:til:) we have configured msgpack encoding for raft to exclusively consider `codec` tags
* Updates the Node drain FSM applier to use msgtype on txn so that drain events are emitted
* Updates the contributing guide with all this, to guide folks in the future on structs-to-api options.

I added [some benchmarking](https://github.com/hashicorp/nomad/pull/10202/files#diff-e950133dfdce118008e42fa60a1930583b929388309ad9e4ba2283b65572ba10R1249-R1272) to see whether the JSON encoding extensions were expensive, with the following results: 
```
BenchmarkHTTPServer_JSONEncodingWithExtensions-12       	   92414	     12141 ns/op
BenchmarkHTTPServer_JSONEncodingWithoutExtensions-12    	   94684	     11569 ns/op
```
So, 600ns cost with the extensions. Do with that what you will. Note, this doesn't compare against doing a struct-to-API conversion as part of encoding (because I didn't want to write it). Also, it's possible that this is slower because of neglected generated structs; I didn't investigate because of https://m.xkcd.com/1319/

### CODE REVIEW

File: nomad/state/events_test.go [hunk 1]
```diff
@@ -120,9 +120,10 @@ func TestEventFromChange_NodeSecretID(t *testing.T) {
 	out := eventsFromChanges(s.db.ReadTxn(), changes)
 	require.Len(t, out.Events, 1)
 
-	nodeEvent, ok := out.Events[0].Payload.(*structs.NodeStreamEvent)
+	_, ok := out.Events[0].Payload.(*structs.NodeStreamEvent)
 	require.True(t, ok)
-	require.Empty(t, nodeEvent.Node.SecretID)
+	// TODO: cgbaker: do we really want to remove this check?
+	// require.Empty(t, nodeEvent.Node.SecretID)
```
Comment (cgbaker) [inline nomad/state/events_test.go:126]: need to find a way to reinsert these tests

File: nomad/state/events_test.go [hunk 2]
```diff
@@ -140,9 +141,10 @@ func TestEventFromChange_NodeSecretID(t *testing.T) {
 	out2 := eventsFromChanges(s.db.ReadTxn(), changes)
 	require.Len(t, out2.Events, 1)
 
-	nodeEvent2, ok := out2.Events[0].Payload.(*structs.NodeStreamEvent)
+	_, ok = out2.Events[0].Payload.(*structs.NodeStreamEvent)
 	require.True(t, ok)
-	require.Empty(t, nodeEvent2.Node.SecretID)
+	// TODO: cgbaker: do we really want to remove this check?
+	// require.Empty(t, nodeEvent2.Node.SecretID)
```
Comment (cgbaker) [inline nomad/state/events_test.go:147]: need to find a way to reinsert these tests

File: nomad/state/events_test.go [hunk 3]
```diff
@@ -94,57 +94,6 @@ func TestEventFromChange_ACLTokenSecretID(t *testing.T) {
 	require.Empty(t, tokenEvent2.ACLToken.SecretID)
 }
 
-// TestEventFromChange_NodeSecretID ensures that a node's secret ID is not
```
Comment (cgbaker) [inline nomad/state/events_test.go:97]: this check is provided in a new test: 
https://github.com/hashicorp/nomad/pull/10202/files#diff-4dd5ab69dbb20720939d5d0edc23108a3b6427943ff1b71f2fea90a79481c80dR168-R169

along with a sibling test for `/v1/node/:id` provided here:
https://github.com/hashicorp/nomad/pull/10202/files#diff-0acc6b10aaabc14cceeb1cbb9b9ded8935451d1f6b4514ee1806b5c9af1ff252R181

File: nomad/stream/ndjson.go [hunk 4]
```diff
@@ -71,15 +74,17 @@ func (n *JsonStream) Send(v interface{}) error {
 		return n.ctx.Err()
 	}
 
-	buf, err := json.Marshal(v)
+	var buf bytes.Buffer
+	enc := codec.NewEncoder(&buf, handlers.JsonHandleWithExtensions)
```
Comment (cgbaker) [inline nomad/stream/ndjson.go:78]: this is so that we just the JSON handler with any registered extensions

File: nomad/structs/structs.go [hunk 5]
```diff
@@ -1965,6 +1940,14 @@ func (n *Node) Canonicalize() {
 			}
 		}
 	}
+
```
Comment (cgbaker) [inline nomad/structs/structs.go:1943]: this is superfluous from the point of view of upgrading... however, there were some tests in the test suite that set `DrainStrategy` but didn't bother to set `SchedulingEligibility`, so this covered for them. also, never pass up a chance to enforce invariants.

File: nomad/structs/structs.go [hunk 6]
```diff
@@ -10599,18 +10582,6 @@ var MsgpackHandle = func() *codec.MsgpackHandle {
 	return h
 }()
 
-var (
```
Comment (cgbaker) [inline nomad/structs/structs.go:10602]: these had to move to another package because of the ordering of init() and module variables and the need to avoid circular imports.

File: api/event_stream_test.go [hunk 7]
```diff
@@ -148,9 +149,24 @@ func TestEventStream_PayloadValue(t *testing.T) {
...
 			n, err := e.Node()
 			require.NoError(t, err)
 			require.NotEqual(t, "", n.ID)
+
+			// raw decoding to verify that the node did not contain SecretID
+			raw := make(map[string]map[string]interface{}, 0)
+			cfg := &mapstructure.DecoderConfig{
+				Result: &raw,
+			}
+
```
Comment (cgbaker) [inline api/event_stream_test.go:162]: doing a raw decoding here into a map to make sure that expected fields are present (i.e., I know how to decode) and unexpected fields ("SecretID") are not

File: api/nodes_test.go [hunk 8]
```diff
@@ -178,6 +178,38 @@ func TestNodes_Info(t *testing.T) {
...
+		if n := len(out); n != 1 {
+			return false, fmt.Errorf("expected 1 node, got: %d", n)
+		}
+		nodeID = out[0].ID
+		return true, nil
+	}, func(err error) {
+		t.Fatalf("err: %s", err)
+	})
+
+	// Query the node, ensure that .SecretID was not returned by the HTTP server
```
Comment (cgbaker) [inline api/nodes_test.go:205]: doing a raw decoding here into a map to make sure that expected fields are present (i.e., I know how to decode) and unexpected fields ("SecretID") are not

File: api/nodes_test.go [hunk 9]
```diff
@@ -218,11 +248,34 @@ func TestNodes_ToggleDrain(t *testing.T) {
 	require.Nil(err)
 	assertWriteMeta(t, &drainOut.WriteMeta)
 
-	// Check again
-	out, _, err = nodes.Info(nodeID, nil)
-	require.Nil(err)
-	if out.SchedulingEligibility != NodeSchedulingIneligible {
-		t.Fatalf("bad eligibility: %v vs %v", out.SchedulingEligibility, NodeSchedulingIneligible)
+	// Drain may have completed before we can check, use event stream
```
Comment (cgbaker) [inline api/nodes_test.go:251]: want to see `Drain` change from true to false

**Review by tgross — APPROVED** [review]
LGTM!

File: api/nodes_test.go [hunk 10]
```diff
@@ -218,11 +248,34 @@ func TestNodes_ToggleDrain(t *testing.T) {
...
+	// we expect to see the node change to Drain:true and then back to Drain:false+ineligible
+	var sawDraining, sawDrainComplete uint64
+	for sawDrainComplete == 0 {
+		select {
+		case events := <-streamCh:
+			require.NoError(events.Err)
+			for _, e := range events.Events {
+				node, err := e.Node()
+				require.NoError(err)
+				if node.Drain && node.SchedulingEligibility == NodeSchedulingIneligible {
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply api/nodes_test.go:269]: certainly won't hurt!

[the remark this answers is not in the document]
Comment (cgbaker) [reply api/nodes_test.go:271]: added this in 4e7d84

File: api/nodes_test.go (hunk 9 above)
Comment (drewbailey) [reply api/nodes_test.go:251]: this is awesome :D

File: nomad/json/encoding.go [hunk 11]
```diff
@@ -0,0 +1,48 @@
+package json
```
Comment (schmichael) [inline nomad/json/encoding.go:1]: Does this package name really not cause conflicts with `encoding/json` all over the place? :sweat_smile: 

Looks like only `nomad/json/handlers` imports it... should we just combine them into a `nomad/jsonhandler` package and unexport all of the current `nomad/json` exports? There may be some import issues your method gets around that I'm missing.

File: nomad/structs/structs.go (hunk 5 above)
Comment (schmichael) [reply nomad/structs/structs.go:1943]: I think this is necessary, and I actually left a question about where it lives on the removal above before scrolling down! :sweat_smile: 

We need to ensure the invariant  `SchedulingEligibility==Ineligible when DrainStrategy!=nil` is enforced on all node updates from the CLI/HTTP-API/RPC. Canonicalize is our usual place of enforcing such constraints, so this seem right to me. Maybe add a comment along the lines of:

> Ensure SchedulingEligibility is set whenever draining so the plan applier and other scheduling logic only need to check SchedulingEligibility when determining whether a placement is feasible on a node.

File: scheduler/util.go [hunk 12]
```diff
@@ -255,7 +255,7 @@ func readyNodesInDCs(state State, dcs []string) ([]*structs.Node, map[string]int
 		if node.Status != structs.NodeStatusReady {
 			continue
 		}
-		if node.Drain {
+		if node.DrainStrategy != nil {
```
Comment (schmichael) [inline scheduler/util.go:258]: Similar to the plan applier logic, we actually don't need to check this at all anymore. Since enforce `SchedulingEligibilty==Ineligible when DrainStategy!=nil` in `Node.Canonicalize()` we can skip this check as we check eligibility below.

**Review by schmichael — CHANGES_REQUESTED** [review]
You deserve a medal for this PR. So many tricky bits to think through! So many pre-0.8 broken tests!

So I _think_ the SecretID bits need to be removed for the reasons I point out in comments. If I'm wrong please tell me because *wow* is reimplementing `SecretID=""` in every single place we read Nodes a huge liability. Removing that liability would be a massive win, but I don't think this method meets our needs.

That being said the Drain bits, and the JSON extensions in particular, look great! I look forward to using them to make future migrations much easier!

File: nomad/json/encoding.go (hunk 11 above)
Comment (cgbaker) [reply nomad/json/encoding.go:1]: excellent point, it caused one or two problems, though not as many as you would expect; most of our code either uses this with go-msgpack or `encoding/json`.

i'll rename it. 👍

File: nomad/node_endpoint.go [hunk 13]
```diff
@@ -811,9 +801,7 @@ func (n *Node) GetNode(args *structs.NodeSpecificRequest,
 
 			// Setup the output
 			if out != nil {
-				// Clear the secret ID
-				reply.Node = out.Copy()
-				reply.Node.SecretID = ""
+				reply.Node = out
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/node_endpoint.go:804]: damn, good catch.
okay, i'll revert this change.

File: nomad/plan_apply.go [hunk 14]
```diff
@@ -648,7 +648,7 @@ func evaluateNodePlan(snap *state.StateSnapshot, plan *structs.Plan, nodeID stri
 		return false, "node is not ready for placements", nil
 	} else if node.SchedulingEligibility == structs.NodeSchedulingIneligible {
 		return false, "node is not eligible for draining", nil
-	} else if node.Drain {
+	} else if node.DrainStrategy != nil {
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/plan_apply.go:651]: `node.Drain` and `node.DrainStrategy != nil` should be equivalent post 0.8.0, correct? so would that scenario only cause problem for upgrades directly from 0.8.0 to 1.1.0?

File: nomad/state/events.go [hunk 15]
```diff
@@ -179,15 +175,11 @@ func eventFromChange(change memdb.Change) (structs.Event, bool) {
...
-		node := after.Copy()
-		node.SecretID = ""
-
 		return structs.Event{
 			Topic: structs.TopicNode,
-			Key:   node.ID,
+			Key:   after.ID,
 			Payload: &structs.NodeStreamEvent{
-				Node: node,
+				Node: after,
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/state/events.go:182]: @schmichael , @tgross , do you think it's worth starting and documenting a `func ... Sanitize()` pattern?

File: nomad/plan_apply.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (schmichael) [reply nomad/plan_apply.go:651]: > node.Drain and node.DrainStrategy != nil should be equivalent post 0.8.0, correct?

Correct! So your direct translation is *not wrong.* In >=0.8 it is merely redundant with the preceeding eligibility check.

> so would that scenario only cause problem for upgrades directly from 0.8.0 to 1.1.0?

Sadly no because even 1.0.x checks `node.Drain` here. So as soon as a server is upgraded to 1.1.0, the 1.0.x servers may improperly see `node.Drain = false` here due to the 1.1.0 server stripping the `true` value when roundtripping the Node struct.

But again! I _think_ we're ok in the `>=0.8 -> 1.1.0` upgrade case because of the SchedulingEligibility invariant and check above!

Upgrading from `<0.8` I'm not worried about. Our recommendation is not skipping versions, and skipping 1 or 2 almost always works fine. Skipping 6+ major releases is... not something I'm interested in supporting at this time. :sweat_smile:

[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/plan_apply.go:651]: thanks, @schmichael , i was missing this nuance, worrying more about the newer versions and less about the older ones.

File: nomad/state/events.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (schmichael) [reply nomad/state/events.go:182]: Yeah, I think the general pattern is something like:

```go
func (t *T) Sanitzie() *T {
  t2 := t.Copy()
  t2.Secrets = ""
  return t2
}
```

Now if we're really clever we'd figure out a way to use different types for these to let the compiler help us assert the right form is going the right place.... but I'm not sure that's tenable. I think the best it could look like is https://play.golang.org/p/33SEO_o4iG1

Which requires changing *a lot* of code, so I don't think that's going to be worth it.

File: nomad/json/extensions.go [hunk 16]
```diff
@@ -0,0 +1,30 @@
...
+	registerExtension(reflect.TypeOf(&structs.Node{}), nodeExt)
+}
+
+// nodeExt adds the legacy field .Drain back to encoded Node objects
+func nodeExt(v interface{}) interface{} {
+	node := v.(*structs.Node)
+	if node == nil {
+		return nil
+	}
+	type NodeAlias structs.Node
```
[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/json/extensions.go:22]: i'll add a comment about what's happening here and clarify the code.
the short answer is, i don't think so. what this is doing is creating a type definition (not an alias, bad naming aside) so that the embedded field has an [identical underlying field](https://golang.org/ref/spec#Conversions); this means that go gives us a free conversion function (for the two types, as well as pointers to them). and because it is a distinct type (unlike an alias), it doesn't recursively/infinitely trigger this encoding extension.

the choice of pointers was just for efficiency, to keep from having to copy the node fields. it's the difference between these two:
pointers: https://play.golang.org/p/WEMAk2YSYBs
no pointers: https://play.golang.org/p/D0oWKEt1lmd

long story short, it says "put a node in here, but pretend it's not a node". i couldn't find another way to do this.

i'll clean this bit up and try to explain the hack/magic.

File: nomad/json/encoding.go (hunk 11 above)
Comment (cgbaker) [reply nomad/json/encoding.go:1]: i remember, i had them in different packages because package variables are instantiated before package init is run, but i'll refactor this to get rid of package init.

File: nomad/plan_apply.go (hunk 14 above)
[the remark this answers is not in the document]
Comment (cgbaker) [reply nomad/plan_apply.go:651]: we should consider these workflows when marking fields for deprecation in the future. adding an `IsDraining() bool` method at the time of deprecation might have eased this.

File: contributing/checklist-jobspec.md [hunk 17]
```diff
@@ -40,3 +42,24 @@ required in the original `jobspec` package.
...
+* convert the struct from the RPC call to the appropriate `api/` struct.
+  This approach is the most developer effort, but it does have a strong guarantee that the HTTP response matches the 
+  API, due to the explicit conversion (assuming proper implementation, which requires tests).
+  Examples of this approach include [GET `/v1/volume/csi/:id`](https://github.com/hashicorp/nomad/blob/v1.0.0/command/agent/csi_endpoint.go#L108)
+* convert to an intermediate struct with the same shape as the `api/` struct.
+  This approach strikes a balance between the former two approaches. 
+  This conversion can be performed in-situ in the agent HTTP handler, as long as the conversion doesn't need to 
+  appear in other handlers. 
+  Otherwise, it is possible to register an extension on the JSON encoding used by the HTTP agent; these extensions
+  can be put in `nomad/json/extensions.go`.
```
Comment (cgbaker) [inline contributing/checklist-jobspec.md:65]:
```suggestion
  can be put in `nomad/jsonhandles/extensions.go`.
```

File: nomad/plan_apply.go [hunk 18]
```diff
@@ -647,10 +647,7 @@ func evaluateNodePlan(snap *state.StateSnapshot, plan *structs.Plan, nodeID stri
 	} else if node.Status != structs.NodeStatusReady {
 		return false, "node is not ready for placements", nil
 	} else if node.SchedulingEligibility == structs.NodeSchedulingIneligible {
-		return false, "node is not eligible for draining", nil
-	} else if node.Drain {
-		// Deprecate in favor of scheduling eligibility and remove post-0.8
-		return false, "node is draining", nil
+		return false, "node is not eligible", nil
```
Comment (cgbaker) [inline nomad/plan_apply.go:650]: changed this error message because it seemed wrong. is that allowed? @schmichael , @tgross

File: nomad/structs/structs.go [hunk 19]
```diff
@@ -1920,26 +1907,36 @@ type Node struct {
...
-	// upgrading nodes
-	if n.SchedulingEligibility == "" {
-		if n.Drain {
-			n.SchedulingEligibility = NodeSchedulingIneligible
-		} else {
-			n.SchedulingEligibility = NodeSchedulingEligible
-		}
+	// Ensure SchedulingEligibility is correctly set whenever draining so the plan applier and other scheduling logic
+	// only need to check SchedulingEligibility when determining whether a placement is feasible on a node.
+	if n.DrainStrategy != nil {
```
Comment (cgbaker) [inline nomad/structs/structs.go:1936]: @schmichael , added a comment as requested and tightened this up to make sure we honor the invariant.

File: scheduler/util.go (hunk 12 above)
Comment (cgbaker) [reply scheduler/util.go]: I updated `Canonicalize` to enforce that

**Comment by cgbaker** [comment]
@schmichael , i rolled back the suggested changes and cleaned up some code and comments. this should be ready to go, if you'll re-review! thanks!

File: nomad/plan_apply.go (hunk 18 above)
Comment (schmichael) [reply nomad/plan_apply.go:650]: Good catch! Your message looks right to me.

**Review by schmichael — APPROVED** [review]
Looks great!
